motionplan/ik/inverseKinematics.go

```diff
@@ -4,30 +4,97 @@ package ik
+	defaultLimit = 999
+
 	// Default distance below which two distances are considered equal.
 	defaultEpsilon = 0.001
 
 	// default amount of closeness to get to the goal.
 	defaultGoalThreshold = defaultEpsilon * defaultEpsilon
 )
 
-// InverseKinematics defines an interface which, provided with seed inputs and a Metric to minimize to zero, will output all found
+// InverseKinematics defines an interface which, provided with seed inputs and a function to minimize to zero, will output all found
 // solutions to the provided channel until cancelled or otherwise completes.
 type InverseKinematics interface {
 	referenceframe.Limited
-	// Solve receives a context, the goal arm position, and current joint angles.
-	Solve(context.Context, chan<- *Solution, []referenceframe.Input, StateMetric, int) error
+	// Solve receives a context, a channel to which solutions will be provided, a function whose output should be minimized, and a
+	// number of iterations to run.
+	Solve(context.Context, chan<- *Solution, []float64, func([]float64) float64, int) error
 }
 
 // Solution is the struct returned from an IK solver. It contains the solution configuration, the score of the solution, and a flag
 // indicating whether that configuration and score met the solution criteria requested by the caller.
 type Solution struct {
-	Configuration []referenceframe.Input
+	Configuration []float64
 	Score         float64
 	Exact         bool
 }
+
+// generateRandomPositions generates a random set of positions within the limits of this solver.
+func generateRandomPositions(randSeed *rand.Rand, lowerBound, upperBound []float64) []float64 {
+	pos := make([]float64, len(lowerBound))
+	for i, l := range lowerBound {
+		u := upperBound[i]
+
+		if l == math.Inf(-1) {
+			l = -defaultLimit
+		}
+		if u == math.Inf(1) {
+			u = defaultLimit
+		}
+
+		jRange := math.Abs(u - l)
+		// Note that rand is unseeded and so will produce the same sequence of floats every time
+		// However, since this will presumably happen at different positions to different joints, this shouldn't matter
+		pos[i] = randSeed.Float64()*jRange + l
+	}
+	return pos
+}
+
+func limitsToArrays(limits []referenceframe.Limit) ([]float64, []float64) {
+	var min, max []float64
+	for _, limit := range limits {
+		min = append(min, limit.Min)
+		max = append(max, limit.Max)
+	}
+	return min, max
+}
+
+// NewMetricMinFunc takes a metric and a frame, and converts to a function able to be minimized with Solve().
+func NewMetricMinFunc(metric StateMetric, frame referenceframe.Frame) func([]float64) float64 {
+	return func(x []float64) float64 {
+		mInput := &State{Frame: frame}
+		inputs := referenceframe.FloatsToInputs(x)
+		eePos, err := frame.Transform(inputs)
+		if eePos == nil || (err != nil && !strings.Contains(err.Error(), referenceframe.OOBErrString)) {
+			globalLogger.Errorw("error calculating frame Transform in IK", "error", err)
+			return 0
```

Comment:
returning zero feels weird here since it corresponds to an error, and we are trying to minimize this function.  Why not some other sentinel value?

Reply:
That's a good point; math.Inf would be better